terminals/Altapay_Test_Terminal.class.php

```diff
@@ -10,401 +10,509 @@
+						$order->update_status( 'failed', 'Payment failed: ' . $errorMessage );
+						$order->add_order_note( __( 'Payment failed: ' . $errorMessage . ' Merchant error: ' . $merchantError, 'altapay' ) );
+					}
+				}
+
+				exit;
+			}
+
+			if ( $status === 'open' ) {
+				$order->update_status( 'on-hold', 'The payment is pending an update from the payment provider.' );
+				$redirect = $this->get_return_url( $order );
+				wp_redirect( $redirect );
+				exit;
+			}
+
+			if ( $payment_status === 'released' ) {
+				$order->add_order_note( __( 'Payment failed: payment released', 'altapay' ) );
+				wc_add_notice( __( 'Payment error:', 'altapay' ) . ' Payment released', 'error' );
+				wp_redirect( wc_get_cart_url() );
+				exit;
+			}
+
+			if ( array_key_exists( 'cancel_order', $_GET ) ) {
+				$order->add_order_note( __( 'Payment failed: ' . $errorMessage . ' Merchant error: ' . $merchantError, 'altapay' ) );
+				wc_add_notice( __( 'Payment error:', 'altapay' ) . ' ' . $errorMessage, 'error' );
+				wp_redirect( wc_get_cart_url() );
+				exit;
+			}
+
+			// Make some validation
+			if ( $errorMessage || $merchantError ) {
+				$order->add_order_note( __( 'Payment failed: ' . $errorMessage . ' Merchant error: ' . $merchantError, 'altapay' ) );
+				wc_add_notice( __( 'Payment error:', 'altapay' ) . ' ' . $errorMessage, 'error' );
+				wp_redirect( wc_get_cart_url() );
+				exit;
+			}
+
+			if ( $order->has_status( 'pending' ) && $status === 'succeeded' ) {
+				// Payment completed
+				$order->add_order_note( __( 'Callback completed', 'altapay' ) );
+				$order->payment_complete();
+				update_post_meta( $order_id, '_transaction_id', $txnId );
+				update_post_meta( $order_id, '_cardno', $cardNo );
+				update_post_meta( $order_id, '_credit_card_token', $credToken );
+				update_post_meta( $order_id, '_credit_card_brand', $creditCardCardBrand );
+				update_post_meta( $order_id, '_credit_card_expiry_date', $creditCardExpiryDate );
+			}
+
+			// Redirect to Order Confirmation Page
+			if ( $type === 'paymentAndCapture' && $requireCapture === 'true' ) {
+				$api = $this->altapayApiLogin();
+				if ( ! $api || is_wp_error( $api ) ) {
+					echo '<p><b>' . __( 'Could not connect to AltaPay!', 'altapay' ) . '</b></p>';
+					return;
+				}
+				$api = new CaptureReservation( $this->getAuth() );
+				$api->setAmount( (float) number_format( $amount, 2, '.', '' ) );
+				$api->setTransaction( $txnId );
+
+				/** @var CaptureReservationResponse $response */
+				try {
+					$response = $api->call();
+				} catch ( ResponseHeaderException $e ) {
+					error_log( 'Response header exception ' . $e->getMessage() );
+				} catch ( \Exception $e ) {
+					error_log( 'Response header exception ' . $e->getMessage() );
+				}
+			}
+			$redirect = $this->get_return_url( $order );
+			wp_redirect( $redirect );
+			exit;
+		}
+	}
+
+	/**
+	 * @param array   $addressInfo
+	 * @param Address $address
+	 *
+	 * @return object
```

Comment:
```suggestion
	 * @return Address
```
Note that it's a bit odd to return on input object as objects are always by reference and so there is no need to also return it.